tests/contrib/pylons/test_pylons.py

```diff
@@ -337,3 +339,100 @@ def test_success_200_ot(self):
         eq_(dd_span.resource, 'root.index')
         eq_(dd_span.meta.get(http.STATUS_CODE), '200')
         eq_(dd_span.error, 0)
+
+    def test_patch(self):
+        # Ensure that patching installs the instrumentation
+        ddtrace.tracer = self.tracer
+
+        patch()
+
+        wsgiapp = loadapp('config:test.ini', relative_to=PylonsTestCase.conf_dir)
+        app = fixture.TestApp(wsgiapp)
+
+        res = app.get(url_for(controller='root', action='index'))
+        eq_(res.status, 200)
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+        span = spans[0]
+
+        eq_(span.service, 'pylons')
+        eq_(span.resource, 'root.index')
+        eq_(span.meta.get(http.STATUS_CODE), '200')
+        eq_(span.error, 0)
+        unpatch()
+
+    def test_unpatch(self):
+        ddtrace.tracer = self.tracer
+
+        patch()
+        unpatch()
+
+        wsgiapp = loadapp('config:test.ini', relative_to=PylonsTestCase.conf_dir)
+        app = fixture.TestApp(wsgiapp)
+
+        res = app.get(url_for(controller='root', action='index'))
+        eq_(res.status, 200)
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 0)
+
+    def test_double_patch(self):
+        ddtrace.tracer = self.tracer
+
+        patch()
+        patch()
+
+        wsgiapp = loadapp('config:test.ini', relative_to=PylonsTestCase.conf_dir)
+        app = fixture.TestApp(wsgiapp)
+
+        res = app.get(url_for(controller='root', action='index'))
+        eq_(res.status, 200)
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+        unpatch()
+
+    def test_double_unpatch(self):
+        ddtrace.tracer = self.tracer
+
+        patch()
+        unpatch()
+        unpatch()
+
+        wsgiapp = loadapp('config:test.ini', relative_to=PylonsTestCase.conf_dir)
+        app = fixture.TestApp(wsgiapp)
+
+        res = app.get(url_for(controller='root', action='index'))
+        eq_(res.status, 200)
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 0)
+
+    def test_patch_auto_and_manual(self):
+        # Test when the app is autoinstrumented and manually instrumented
+        ddtrace.tracer = self.tracer
+
+        patch()
+
+        wsgiapp = loadapp('config:test.ini', relative_to=PylonsTestCase.conf_dir)
+        app = PylonsTraceMiddleware(wsgiapp, self.tracer, service='manual')
+        app = fixture.TestApp(app)
+
+        res = app.get(url_for(controller='root', action='index'))
+        eq_(res.status, 200)
+
+        spans = self.tracer.writer.pop()
+
+        # There should be two spans one from each instrumentation
+        eq_(len(spans), 2)
```

Comment:
Note this behaviour. Typically we have tried to avoid duplicating spans when a user manually instruments and when `ddtrace-run/patch()/patch_all()` are used.

There isn't much we can do to prevent the manual installation of the tracing middleware in pylons. We could maybe attach some marker to the request in the manual middleware that can be checked for in the automatic middleware.

However, having 2 spans might be desirable as now users will gain more insight into the timing of the middleware.

Reply:
Cant we do a check somewhere for `if isinstance(app, TracedPylonsApp): # skip patching` ?

Reply:
Hmm. I think we would like to opt to keep the manual instrumentation over the auto instrumentation which makes things trickier.

Reply:
I did this in Flask, basically you can keep the manual middleware around for people who import it and instrument manually, but if they are just using `patch()` we can change the behavior since the API isn't changing

Reply:
We talked about this offline, we'll keep the manual and the auto together.

It will double patch, but that is ok, it is visibility for the user, and ensures nothing breaks.